src/audio/host.c

```diff
@@ -350,6 +351,14 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
 	if (!hd->no_stream_position) {
 		hd->report_pos += bytes;
 
+		/* update position every period to support host query to
+		 * get the latest value if cont_update_posn == 1.
+		 */
+		if (hd->cont_update_posn == 1) {
+			pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+			mailbox_stream_write(dev->pipeline->posn_offset,
+					     &hd->posn, sizeof(hd->posn));
+		}
```

Comment:
The `hd->report_pos` is not used if `hd->cont_update_posn == 1`

```c
if (!hd->no_stream_position) {
	hd->report_pos += bytes;

	if (hd->cont_update_posn == 1) {
		/*
		 * update position every period to support host query to get the
		 * latest value if cont_update_posn == 1.
		 */
		pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
		mailbox_stream_write(dev->pipeline->posn_offset,
					&hd->posn, sizeof(hd->posn));
	} else if (hd->host_period_bytes != 0 &&
		   hd->report_pos >= hd->host_period_bytes) {
		/*
		 * host_period_bytes is set to zero to disable position update
		 * by IPC for FW version before 3.11, so send IPC message to
		 * driver according to this condition and report_pos.
		 */
		hd->report_pos = 0;
		/* send timestamped position to host
		* (updates position first, by calling ops.position())
		*/
		pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
		mailbox_stream_write(dev->pipeline->posn_offset,
				&hd->posn, sizeof(hd->posn));
		ipc_msg_send(hd->msg, &hd->posn, false);
	}
}
```
But I would be tempted to move the `hd->report_pos += bytes;` to the else case as with continuous update you do not even need to increment it.

What would be the side effect if this is done unconditionally, without a flag from the host?

Reply:
@ujfalusi  hd->report_pos is still needed when hd->cont_update_posn == 1 because when hd->cont_update_posn is set as '1', only update host position to mailbox buffer without sending ipc to notify the host.
About the side effect without a flag from the host,  
Please reference the discussion as the link https://github.com/thesofproject/sof/pull/5137#discussion_r776066511. @lgirdwood has the suggestion about it. Thanks

Reply:
@yaochunhung, I see. so this will update the position data in between IPCs, `no_stream_position == 0` -> the position IPC is enabled.
How is the position updated in case of `no_stream_position == 1'?

Reply:
Please see my review comment, I think we should clean up the semantics, the old fields are quite hard to follow (and worse if we add more fields on top). I think host should still have option to prevent both mailbox and IPC updates (traffic to mailbox is non-zero cost on many platforms so if host can get position info via other means, it probably doesn't need the mailbox traffic -> the case on Intel platforms)